lib/Sema/TypeOfReference.cpp

```diff
@@ -2261,10 +2261,11 @@ static DeclReferenceType getTypeOfReferenceWithSpecialTypeCheckingSemantics(
         CS.getConstraintLocator(locator, ConstraintLocator::ThrownErrorType),
         0);
     FunctionType::Param arg(escapeClosure);
+    bool isAsync = CS.isAsynchronousContext(useDC);
     auto bodyClosure = FunctionType::get(arg, result,
                                          FunctionType::ExtInfoBuilder()
                                              .withNoEscape(true)
-                                             .withAsync(true)
+                                             .withAsync(isAsync)
```

Comment:
Looks like this was originally added in https://github.com/swiftlang/swift/pull/35664